### PR TITLE
Initialize grad before Optimizer calls update method

### LINKED ISSUE #1786: Unused params cause NoneType error
When some params in a model are not used to calculate loss value, its gradient is None. It causes an error when an optimizer update its value.

```
import numpy

import chainer
import chainer.functions as F
import chainer.links as L
import chainer.optimizers as O


class MyModel(chainer.Chain):

    def __init__(self):
        super(MyModel, self).__init__(
            l1=L.Linear(3, 4),
            l2=L.Linear(3, 4),
        )

    def __call__(self, x):
        return F.sum(self.l1(x))


model = MyModel()
opt = O.Adam()
opt.setup(model)
model.cleargrads()
loss = model(numpy.random.uniform(-1, 1, (5, 3)).astype('f'))
loss.backward()
opt.update()
```

### PR DESCRIPTION
**Pull request by unnonouno**
I fixed optimizer to initialize grads of parameters before calling hooks and update method. This fix is ineffective when updater does not affect param value when grad is None. I think efficiency is another issue.

fix #1786

### CODE REVIEW

**Comment by okuta** [comment]
LGTM!
